File: ios/Classes/GatiFlowPlugin.swift
import Flutter
import UIKit
import GatiFlow

public class GatiFlowPlugin: NSObject, FlutterPlugin {

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "dev.gatiflow/sdk",
            binaryMessenger: registrar.messenger()
        )
        let instance = GatiFlowPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any] ?? [:]

        switch call.method {

        // ── Initialization ─────────────────────────────────────────────────
        case "start":
            guard let appToken = args["appToken"] as? String else {
                return result(FlutterError(code: "MISSING_TOKEN", message: "appToken is required", details: nil))
            }
            let services   = args["services"]          as? [String] ?? ["crashes", "analytics"]
            let baseUrl    = args["baseUrl"]            as? String   ?? "https://app.gatiflow.dev"
            let debug      = args["debugLogging"]       as? Bool     ?? false
            let crashSize  = args["maxCrashQueueSize"]  as? Int      ?? 50
            let batchSize  = args["maxEventBatchSize"]  as? Int      ?? 100
            let flushMs    = args["flushIntervalMs"]    as? Double   ?? 30_000
            let watchdogMs = args["anrTimeoutMs"]       as? Double   ?? 5_000

            let config = Config.Builder(appToken: appToken)
                .baseUrl(baseUrl)
                .debugLogging(debug)
                .maxCrashQueueSize(crashSize)
                .maxEventBatchSize(batchSize)
                .flushIntervalMs(flushMs)
                .watchdogTimeoutMs(watchdogMs)
                .build()

            let serviceList: [GatiFlowService] = services.compactMap {
                switch $0 {
                case "crashes":   return Crashes()
                case "analytics": return Analytics()
                default:          return nil
                }
            }

            GatiFlow.shared.start(config: config, services: serviceList)
            result(nil)

        // ── Analytics ──────────────────────────────────────────────────────
        case "trackEvent":
            let name  = args["name"]       as? String         ?? ""
            let props = args["properties"] as? [String: Any]
            GatiFlow.shared.analytics?.trackEvent(name, properties: props)
            result(nil)

        case "flush":
            GatiFlow.shared.analytics?.flush()
            result(nil)

        case "setAnalyticsEnabled":
            GatiFlow.shared.analytics?.setEnabled(args["enabled"] as? Bool ?? true)
            result(nil)

        // ── Crashes ────────────────────────────────────────────────────────
        case "trackError":
            let message  = args["message"]  as? String            ?? ""
            let metadata = args["metadata"] as? [String: String]  ?? [:]
            struct BridgedError: Error, LocalizedError {
                let msg: String
                var errorDescription: String? { msg }
            }
            GatiFlow.shared.crashes?.trackError(BridgedError(msg: message), metadata: metadata)
            result(nil)

        case "setCrashesEnabled":
            GatiFlow.shared.crashes?.setEnabled(args["enabled"] as? Bool ?? true)
            result(nil)

        // ── Identity & lifecycle ───────────────────────────────────────────
        case "setUserId":
            GatiFlow.shared.setUserId(args["userId"] as? String)
            result(nil)

        case "stop":
            GatiFlow.shared.stop()
            result(nil)

        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

File: android/src/main/kotlin/dev/gatiflow/flutter/GatiFlowPlugin.kt
package dev.gatiflow.flutter

import android.app.Application
import dev.gatiflow.sdk.GatiFlow
import dev.gatiflow.sdk.analytics.Analytics
import dev.gatiflow.sdk.core.Config
import dev.gatiflow.sdk.crashes.Crashes
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

class GatiFlowPlugin : FlutterPlugin, MethodCallHandler {

    private lateinit var channel: MethodChannel
    private var application: Application? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, "dev.gatiflow/sdk")
        channel.setMethodCallHandler(this)
        application = binding.applicationContext as? Application
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {

            // ── Initialization ─────────────────────────────────────────────
            "start" -> {
                val appToken   = call.argument<String>("appToken") ?: return result.error("MISSING_TOKEN", "appToken is required", null)
                val services   = call.argument<List<String>>("services") ?: listOf("crashes", "analytics")
                val baseUrl    = call.argument<String>("baseUrl") ?: "https://app.gatiflow.dev"
                val debug      = call.argument<Boolean>("debugLogging") ?: false
                val crashSize  = call.argument<Int>("maxCrashQueueSize") ?: 50
                val batchSize  = call.argument<Int>("maxEventBatchSize") ?: 100
                val flushMs    = call.argument<Int>("flushIntervalMs")?.toLong() ?: 30_000L
                val anrMs      = call.argument<Int>("anrTimeoutMs")?.toLong() ?: 5_000L

                val config = Config.Builder(appToken)
                    .baseUrl(baseUrl)
                    .debugLogging(debug)
                    .maxCrashQueueSize(crashSize)
                    .maxEventBatchSize(batchSize)
                    .flushIntervalMs(flushMs)
                    .anrTimeoutMs(anrMs)
                    .build()

                val serviceClasses = buildList {
                    if ("crashes"   in services) add(Crashes::class.java)
                    if ("analytics" in services) add(Analytics::class.java)
                }

                application?.let { GatiFlow.start(it, config, serviceClasses) }
                result.success(null)
            }

            // ── Analytics ──────────────────────────────────────────────────
            "trackEvent" -> {
                val name  = call.argument<String>("name") ?: return result.error("MISSING_NAME", "event name is required", null)
                val props = call.argument<Map<String, Any>>("properties") ?: emptyMap()
                GatiFlow.analytics?.trackEvent(name, props)
                result.success(null)
            }
            "flush" -> {
                GatiFlow.analytics?.flush()
                result.success(null)
            }
            "setAnalyticsEnabled" -> {
                GatiFlow.analytics?.setEnabled(call.argument<Boolean>("enabled") ?: true)
                result.success(null)
            }

            // ── Crashes ────────────────────────────────────────────────────
            "trackError" -> {
                val message  = call.argument<String>("message") ?: ""
                val metadata = call.argument<Map<String, String>>("metadata") ?: emptyMap()
                GatiFlow.crashes?.trackError(RuntimeException(message), metadata)
                result.success(null)
            }
            "setCrashesEnabled" -> {
                GatiFlow.crashes?.setEnabled(call.argument<Boolean>("enabled") ?: true)
                result.success(null)
            }

            // ── Identity & lifecycle ───────────────────────────────────────
            "setUserId" -> {
                GatiFlow.setUserId(call.argument<String?>("userId"))
                result.success(null)
            }
            "stop" -> {
                GatiFlow.stop()
                result.success(null)
            }

            else -> result.notImplemented()
        }
    }
}
